luigi/contrib/pai.py

```diff
@@ -0,0 +1,300 @@
+        'name', 'taskNumber', 'cpuNumber', 'memoryMB', 'shmMB', 'gpuNumber', 'portList', 'command',
+        'minFailedTaskCount', 'minSucceededTaskCount'
+    )
+
+    def __init__(self, name, command, taskNumber=1, cpuNumber=1, memoryMB=2048, shmMB=64, gpuNumber=0, portList=[]):
+        """
+        The TaskRole of PAI
+
+        :param name: Name for the task role, need to be unique with other roles, required
+        :param command: Executable command for tasks in the task role, can not be empty, required
+        :param taskNumber: Number of tasks for the task role, no less than 1, required
+        :param cpuNumber: CPU number for one task in the task role, no less than 1, required
+        :param shmMB: Shared memory for one task in the task role, no more than memory size, required
+        :param memoryMB: Memory for one task in the task role, no less than 100, required
+        :param gpuNumber: GPU number for one task in the task role, no less than 0, required
+        :param portList: List of portType to use, optional
+        """
+        self.name = name
+        self.command = command
+        self.taskNumber = taskNumber
+        self.cpuNumber = cpuNumber
+        self.memoryMB = memoryMB
+        self.shmMB = shmMB
+        self.gpuNumber = gpuNumber
+        self.portList = portList
+
+
+class OpenPai(luigi.Config):
+    pai_url = luigi.Parameter(
+        default='http://127.0.0.1:9186',
+        description='rest server url, default is http://127.0.0.1:9186')
+    username = luigi.Parameter(
+        default='admin',
+        description='your username')
+    password = luigi.Parameter(
+        default=None,
+        description='your password')
+    expiration = luigi.Parameter(
+        default='3600',
+        description='expiration time in seconds')
+
+
+class PaiTask(luigi.Task):
+    __POLL_TIME = 5
+
+    @abc.abstractproperty
+    def name(self):
+        """Name for the job, need to be unique, required"""
+        return 'SklearnExample'
+
+    @abc.abstractproperty
+    def image(self):
+        """URL pointing to the Docker image for all tasks in the job, required"""
+        return 'openpai/pai.example.sklearn'
+
+    @abc.abstractproperty
+    def tasks(self):
+        """List of taskRole, one task role at least, required"""
+        return []
+
+    @property
+    def auth_file_path(self):
+        """Docker registry authentication file existing on HDFS, optional"""
+        return None
+
+    @property
+    def data_dir(self):
+        """Data directory existing on HDFS, optional"""
+        return None
+
+    @property
+    def code_dir(self):
+        """Code directory existing on HDFS, should not contain any data and should be less than 200MB, optional"""
+        return None
+
+    @property
+    def output_dir(self):
+        """Output directory on HDFS, $PAI_DEFAULT_FS_URI/Output/$jobName will be used if not specified, optional"""
+        return '$PAI_DEFAULT_FS_URI/{0}/output'.format(self.name)
```

Comment:
The docstring and this value don't match

Reply:
Will correct the docstring.